BASH GitHub CLI version of High Resolution cases used for CI weekly tests

TerrenceMcGuinness-NOAA opened the pull request:
<!-- PLEASE READ -->
**Description**

A BASH script using  **GitHub CLI** is added to create a branch from _develop_ named _weekly_tests_ and moves the cases from `$HOMEgfs/ci/cases/weekly` to `$HOMEgfs/ci/cases/pr` and then opens a PR and labels it with _CI-Hera-Ready_ and _CI-Orion-Ready_ thus creating a weekly high resolution test run that can easily be launched by a **cron** job.

Resolves #1663 

**Type of change**
- New feature (adds functionality)
- Maintenance (new CI test)

**Tests Ran**
- Ran `weekly_tests.sh` and successfully create PR with labels and new cases in pr directory

**Checklist**
- [ ] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [ ] I have made corresponding changes to the documentation if necessary

TerrenceMcGuinness-NOAA committed:
added updated weekly case files to forked develop for testing when cloning

TerrenceMcGuinness-NOAA committed:
updated idate to 2023040200 to match files staged in ICSDIRS for both high resolution C384 cases

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

TerrenceMcGuinness-NOAA committed:
Merge branch 'develop' of github.com:TerrenceMcGuinness-NOAA/global-workflow into develop

TerrenceMcGuinness-NOAA committed:
added PyGithub version of weekly tests

TerrenceMcGuinness-NOAA committed:
added bash pr based weekly script

TerrenceMcGuinness-NOAA committed:
added bash version of pr based weekly tests

TerrenceMcGuinness-NOAA committed:
updated bash linting for variable quoting

WalterKolczynski-NOAA commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+HOMEgfs="$(cd "$(dirname  "${BASH_SOURCE[0]}")/../.." >/dev/null 2>&1 && pwd )"
+scriptname=$(basename "${BASH_SOURCE[0]}")
+echo "Begin ${scriptname} at $(date -u)" || true
+export PS4='+ $(basename ${BASH_SOURCE})[${LINENO}]'
```
```suggestion
export PS4='+ $(basename ${BASH_SOURCE[0]})[${LINENO}]'
```

WalterKolczynski-NOAA commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+export PS4='+ $(basename ${BASH_SOURCE})[${LINENO}]'
+
+#########################################################################
+#  Set up runtime environment varibles for accounts on supproted machines
```
```suggestion
# Set up runtime environment variables for accounts on supported machines
```

WalterKolczynski-NOAA commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+export REPO_URL=${REPO_URL:-"https://github.com/NOAA-EMC/global-workflow.git"}
+
+################################################################
+# Setup the reletive paths to scripts and PS4 for better logging 
```
```suggestion
# Setup the relative paths to scripts and PS4 for better logging
```

WalterKolczynski-NOAA commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+#
+# Abstract:
+#
+# This script runs the high resolution cases found in $HOMEgfs/ci/cases/weekly
+# from the develop branch for the global-workflow repo that is intended to run on a weekly basis
+# from a cron job. When ran it will clone and build a new branch from the EMC's global-workflow and
+# and create a pr using GitHub CLI by moving and replacing the yaml case files in
+# ${HOMEgfs}/ci/cases/weekly to {HOMEgfs}/ci/cases/pr.  Then at point it can simply also add the
+# the requisite labels so that current BASH CI framework can run these cases.
```
```suggestion
# This script runs the high resolution cases found in $HOMEgfs/ci/cases/weekly
# from the develop branch for the global-workflow repo that are intended to run on a weekly basis
# from a cron job. When run it will clone and build a new branch from the EMC's global-workflow and
# and create a pr using GitHub CLI by moving and replacing the yaml case files in
# ${HOMEgfs}/ci/cases/weekly to {HOMEgfs}/ci/cases/pr.  Then at point it can simply also add the
# the requisite labels so that current BASH CI framework can run these cases.
```

WalterKolczynski-NOAA commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+# and create a pr using GitHub CLI by moving and replacing the yaml case files in
+# ${HOMEgfs}/ci/cases/weekly to {HOMEgfs}/ci/cases/pr.  Then at point it can simply also add the
+# the requisite labels so that current BASH CI framework can run these cases.
+#######################################################################################i######
```
```suggestion
##############################################################################################

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+#!/bin/bash
+set -eux
+
+##############################################################################################
```
I just want to make sure we note that this script only needs to run on a single machine and not on every machine the CI is being run on.  This is so because all this script does is create a branch push it to the users' fork and apply CI triggering labels.

TerrenceMcGuinness-NOAA replied:
Done.  It's weird that ChatGPT knew enough about the inner workings of the script to finish my sentences  when I was adding this into the comment section.

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+for label in "${PULL_REQUEST_LABELS[@]}"
+do
+  "${GH}" pr edit --add-label "${label}"
+done
```
At this point, I think you can cd out and remove the local clone.
It is no longer necessary.

TerrenceMcGuinness-NOAA replied:
Agreed.

TerrenceMcGuinness-NOAA replied:
Done

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+REPO_NAME="global-workflow"
+BASE_BRANCH="develop"
+HEAD_BRANCH="${branch}"
+PULL_REQUEST_TITLE="Weekly High Resolution Forecast Tests $(date +'%A %b %Y')"
```
```suggestion
PULL_REQUEST_TITLE="[DO NOT MERGE] Weekly High Resolution CI Tests $(date +'%A %b %Y')"
```

TerrenceMcGuinness-NOAA replied:
Nice

WalterKolczynski-NOAA commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+
+##############################################################################################
+#
+# Script description: Top level driver script for checking PR
+#                     ready for CI regression testing
```
```suggestion
# Script description: Top level driver script for running
#                     weekly CI regression tests
```

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+# Create Pull Request using GitHub CLI and add labels for CI testing
+####################################################################
+
+REPO_OWNER="TerrenceMcGuinness-NOAA"
```
Do we plan to use @emcbot for this purpose at some point??

TerrenceMcGuinness-NOAA replied:
Yes, I thought I would test that once its in the develop branch.

WalterKolczynski-NOAA reviewed (changes requested):
Just one minor syntax issue, plus fixing the comments.

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+#########################################################
+# Create a new branch from develop and move yaml files
+#########################################################
+branch="weekly_ci_$(date +%A_%b_%Y)"
```
`date +%A_%b_%Y` yeilds `WeekDayName_Mon_YYYY`.  It should contain `day` as `DD`.  I suggest the form YYYY-MM-DD 
```suggestion
branch="weekly_ci_$(date +%Y%m%d)"
```

TerrenceMcGuinness-NOAA replied:
Personally I like the words, but ok.

aerorahul replied:
Words are fine, but remember this is weekly test, so the branch for all 4 weeks (assume Friday) of September 2023 will be:
`Friday_Sept_2023`.
It does not tell you which specific date it is: Sept 7, 14, 21 or 28.

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+develop_dir="${GFS_CI_ROOT}/develop_temp"
+HOMEgfs_PR="${develop_dir}/global-workflow"
+echo "Creating new branch ${branch} from develop on ${MACHINE_ID} in ${develop_dir}"
+export HOMEgfs_PR
```
```suggestion
```
Not necessary to export this.

TerrenceMcGuinness-NOAA replied:
Oh that's right because the setupexpt.py is not ran here yet.

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,104 @@
+mv "${HOMEgfs_PR}/ci/cases/weekly" "${HOMEgfs_PR}/ci/cases/pr"
+cd "${HOMEgfs_PR}" || exit 1
+git add "${HOMEgfs_PR}/ci/cases/pr"
+git commit -m "Moved weekly cases files into pr for high resultuion testing"
```
```suggestion
git commit -m "Moved weekly cases files into pr for high resolution testing"
```

aerorahul reviewed (changes requested):
Looks great.  
Some minor updates to clarify the branch name and cleaning up after the script finishes.

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,103 @@
+REPO_NAME="global-workflow"
+BASE_BRANCH="develop"
+HEAD_BRANCH="${branch}"
+PULL_REQUEST_TITLE="[DO NOT MERGE] Weekly High Resolution CI Tests $(date +'%A %b %Y')"
```
```suggestion
PULL_REQUEST_TITLE="[DO NOT MERGE] Weekly High Resolution CI Tests $(date +'%A %b %d, %Y')"
```
Add what date the test is from.

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,103 @@
+#########################################################################
+
+source "${HOMEgfs}/ush/detect_machine.sh"
+case ${MACHINE_ID} in
+  hera | orion)
+    echo "Running Automated Testing on ${MACHINE_ID}"
+    source "${HOMEgfs}/ci/platforms/${MACHINE_ID}.sh"
+    ;;
+  *)
+    echo "Unsupported platform. Exiting with error."
+    exit 1
+    ;;
+esac
```
Can you explain why this block is required?

TerrenceMcGuinness-NOAA replied:
We still need a common GFS_CI_ROOT accross all platforms so we can run it that one time anywhere

aerorahul replied:
Ok.  I see `GFS_CI_ROOT` is used in this script to define the location of `develop_temp`.  That location could as well be in `/tmp`, but ok!

TerrenceMcGuinness-NOAA committed:
added more comments so show only needs to run on a single machine and remove temp files for the git push

TerrenceMcGuinness-NOAA committed:
Update driver_weekly.sh

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver_weekly.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA reviewed (commented):
Almost ready, left comment about sourcing platform env for the temp space GFS_CI_ROOT.

TerrenceMcGuinness-NOAA reviewed (commented):
leave a comment for commit

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,106 @@
+
+rm -Rf "${HOMEgfs_PR}/ci/cases/pr"
+mv "${HOMEgfs_PR}/ci/cases/weekly" "${HOMEgfs_PR}/ci/cases/pr"
+cd "${HOMEgfs_PR}" || exit 1
```
I think we are already in `${HOMEgfs_PR}`.  There is no need to cd into it again.

TerrenceMcGuinness-NOAA committed:
removed HOMEgfs_PR and full paths for readablity

TerrenceMcGuinness-NOAA committed:
Merge branch 'feature/bash_weekly_pr_CI' of github.com:TerrenceMcGuinness-NOAA/global-workflow into feature/bash_weekly_pr_CI

aerorahul commented on ci/scripts/driver_weekly.sh:
```diff
@@ -0,0 +1,103 @@
+do
+  "${GH}" pr edit --add-label "${label}"
+done
+rm -Rf "${develop_dir}"
```
The script may have trouble removing this directory as it is presently in that CWD/global-workflow.

TerrenceMcGuinness-NOAA replied:
Oh right, it should work because the shell quits on return, but yeah it's not a good idea.

TerrenceMcGuinness-NOAA committed:
Update driver_weekly.sh

move off of directory that is being removed at the end of the driver_weekly bash script

TerrenceMcGuinness-NOAA reviewed (commented):
updated code for suggestion about deleting dir the shell is in

aerorahul reviewed (approved):
Looks good.  Lets get this to run weekly.
